### PR TITLE
feat: core algorithm with concurrency

### PR DESCRIPTION
**Pull request by moreirathomas**
## Description

Basic implementation of the core algorithm. The runner is the perfect candidate to use concurrent programming (it is a technical requirement for the project).

As of now, this is achieved with the semaphore pattern as used in the `request` package.

## Changes

`request` package regroups code to make a HTTP request (GET only for this POC) and record its details (for now status code, time taken and response body length).
Refer to `request.Do` and `request.DoUntil` documentation for initial specs regarding the concurrent algorithm itself.

It also provide an helper function to collect the results from `request.Do` and `request.DoUntil` as they are send into an out channel. I anticipate this may become its own package (`report`?) once the processing grows in details regarding what will be send to the api server (outside the scope of this PR).

Current usage:

```
Usage of runner:
  -c int
        Number of connections to run concurrently (default 1)
  -d int
        Duration of test, in seconds (default 60)
  -r int
        Number of requests to run, use duration as exit condition if omitted
  -t int
        Timeout for each http request, in seconds (default 10)
```
## Notes

I would like to explore other concurrency patterns before settling on the correct one (or at least the one we will keep).

**Edit: I still think exploring other concurrency patterns would be nice, however merging with this implementation is a better choice, as we can build upon this for other parts of the app. We may re-open an issue regarding other patterns later.**

### CODE REVIEW

File: request/request.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (r Record) String() string {
+	return fmt.Sprintf("status %s, took %s", r.code, r.cost)
+}
+
+func Do(url string, timeout time.Duration) Record {
+	client := http.Client{
+		// Timeout includes connection time, any redirects, and reading the response body.
+		// We may want exclude reading the response body in our benchmark tool.
```
Comment (moreirathomas) [inline request/request.go:56]: See this [answer](https://stackoverflow.com/a/16930649) for "How to set timeout for http.Get() requests in Golang?".

Comment (moreirathomas) [reply request/request.go]: This might become its own issue for later ?

**Review by GregoryAlbouy — CHANGES_REQUESTED** [review]
Ok that semaphore pattern is neat 🔥 
I made some suggestions and required one change.

Also, given the criticity of this part of the app we definitely need unit tests for `Do` and `DoUntil` (they would have catched the issue I mention in the comments). Maybe in another PR (ideally not) but it must be done as soon as possible.

At some point we'll also need some input validation: out of scope here, so I suggest we make the specs and update [#2 Parse input into data structure for core](https://github.com/benchttp/runner/issues/2) accordingly..

File: request/do.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	t0 := time.Now()
+
+	resp, err := client.Get(url)
+	if err != nil {
+		return Record{error: fmt.Sprint(err)}
+	}
+
+	return newRecord(resp, time.Since(t0))
```
Comment (GregoryAlbouy) [inline request/do.go:24]: We shouldn't stop the timer at the end of the function, but rather as soon as the request is over so we don't include our own error handling in it (and potentially more if this function gets updated) :
```suggestion
	t0 := time.Now()
	
	resp, err := client.Get(url)
	elasped := time.Since(t0)
	if err != nil {
		return Record{error: fmt.Sprint(err)}
	}

	return newRecord(resp, elapsed)
```

File: request/request.go (hunk 1 above)
Comment (GregoryAlbouy) [reply request/request.go]: I'm wondering whether we should include it or not, it could be a valid indicator of an endpoint's performance: if it's super quick but delivers a super long body to read we could miss that information 🤔 Could also be an option.

Worth discussing indeed!

Comment (moreirathomas) [reply request/request.go]: Yes it can go both ways. This need more attention and discussion.
Let's not bother with it for now, open an issue and tag it for consideration? So we don't block this PR on that.

File: cmd/runner/main.go [hunk 3]
```diff
@@ -1,7 +1,72 @@
...
+
+	"github.com/benchttp/runner/report"
+	"github.com/benchttp/runner/request"
+)
+
+const (
+	DefaultConcurrency = 1
+	DefaultRequests    = 0 // Use duration as exit condition if omitted.
+	DefaultDuration    = 60
+	DefaultTimeout     = 10
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply cmd/runner/main.go:17]: My thought on this was: it will appear in godocs as it is public. Which can be cool,  because it provide useful information to someone who would use this.

File: request/do.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	rec := make(chan Record, requests)
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < requests; i++ {
+		select {
+		case <-quit:
+			break
+		default:
+		}
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/do.go:59]: Are you sure? A quick manual test worked. Though I might have missed things.

File: request/record.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		code = "1xx"
+	case 2:
+		code = "2xx"
+	case 3:
+		code = "3xx"
+	case 4:
+		code = "4xx"
+	case 5:
+		code = "5xx"
+	}
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/record.go:31]: Nice ! But way less explicit ! You're right anyway I'll keep the exact status for the api server.

File: request/do.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply request/do.go:59]: Yup, a unit test would prove it easily, but you can check it this way:
```go
for i := 0; i < requests; i++ {
	fmt.Println(i)
	select {
	case <-quit:
		fmt.Println("STOP")
		break
	default:
	}
	// ...
}
```

```
> go run cmd/runner/main.go -c 1 -d 1 -r 10 http://localhost:9999\?delay\=500
Testing url: http://localhost:9999?delay=500
concurrency: 1
requests: 10
duration: 1s

0
1
2
3
STOP
4
5
6
7
8
9
10
```

File: cmd/runner/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply cmd/runner/main.go:17]: Hah good point! Makes sense, although I'm surprised it actually doesn't appear 🤔 It seems the docs from `cmd` are ignored anyway.

File: report/report.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+import "github.com/benchttp/runner/request"
+
+func Collect(c <-chan request.Record) []request.Record {
+	rec := []request.Record{}
+
+	for r := range c {
+		rec = append(rec, r)
+	}
+	return rec
+}
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply report/report.go]: Yeah you're right. I saw future use and split the two, but it can wait for when the need arises.

**Comment by moreirathomas** [comment]
Agreed on the need to write unit tests but as it deals with concurrency, goroutines, channels and maximum threads at once, I'm not sure how to do that right now. This needs more work, maybe in a later PR?

File: request/do.go [hunk 7]
```diff
@@ -52,21 +52,23 @@ func Do(requests int, quit <-chan struct{}, concurrency int, url string, timeout
...
-		default:
+	go func() {
+		for i := 0; i < requests; i++ {
+			select {
+			case <-quit:
+				wg.Wait()
+				close(rec)
+				return
+			default:
+			}
```
Comment (GregoryAlbouy) [inline request/do.go:63]: [Bug] The main problem is fixed, but we still have an issue here.
Given that modification:
```go
for i := 0; i < requests; i++ {
	fmt.Println(i)
	select {
	case <-quit:
		fmt.Println("STOP")
		wg.Wait()
		close(rec)
		return
	default:
	}
	// ...
}
```

and this call:
```sh
> go run cmd/runner/main.go -c 1 -d 1s -r 10 http://localhost:9999\?delay\=500
```

I expect something like:

```txt
0
1
STOP
```

but got instead:

```txt
0
1
2
3
STOP
3
```

So I see 2 problems:
- Too many requests done before STOP
- A duplicated 3 after the STOP

**Review by GregoryAlbouy — COMMENTED** [review]
👌 
Last things before merging:
- [ ] Address 2 remaining bugs (weird but probably not critical): either fix them or create an issue
- [ ] Create the PR for unit tests
- [ ] Create discussion about measuring the reading of the request body
- [ ] Create discussion about input validation and link it to #2

File: request/do.go (hunk 7 above)
Comment (moreirathomas) [reply request/do.go:63]: Last `3` comes from `println(len(rec))` in main.

For the "too many requests": I get what you mean, but i don't really know what's going on actually 😅

File: request/do.go [hunk 8]
```diff
@@ -52,21 +52,23 @@ func Do(requests int, quit <-chan struct{}, concurrency int, url string, timeout
 
 	var wg sync.WaitGroup
 
-	for i := 0; i < requests; i++ {
-		select {
-		case <-quit:
-			break
-		default:
+	go func() {
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/do.go:55]: goroutine is needed here, so we can in parallel collect records as they come in at line 72 (`collect(rec)`).

File: request/do.go (hunk 7 above)
Comment (GregoryAlbouy) [reply request/do.go:63]: > Last `3` comes from `println(len(rec))` in main (number of results). First `3` is printed before the `select` statement.

🤦‍♂️ 

As for the too many request, for some reason the first two seem to pop ± simultaneously. I think we can live with that for now, it'll need to be fixed in the PR for the unit test anyway.

Comment (moreirathomas) [reply request/do.go:63]: yep !

File: cmd/runner/main.go [hunk 9]
```diff
@@ -47,19 +48,14 @@ func parseArgs() {
...
+	ctx, cancel := context.WithTimeout(context.Background(), duration)
+	defer cancel()
 
 	var rec <-chan request.Record
 	if requests > 0 {
-		rec = request.Do(requests, quit, concurrency, url, timeout)
+		rec = request.Do(requests, ctx.Done(), concurrency, url, timeout)
 	} else {
-		rec = request.DoUntil(quit, concurrency, url, timeout)
+		rec = request.DoUntil(ctx.Done(), concurrency, url, timeout)
```
[the remark this answers is not in the document]
Comment (moreirathomas) [reply cmd/runner/main.go:57]: For clarification, what do you suggest: 
```go
// Always pass a context.Contex (emphasis on the type) as first arg
func Do(ctx context.Context, ...args) {
	select{
	case <-ctx.Done():
	} 
}
Do(ctx, args...)
```
```go
Always pass quit signal (whatever it is) as first arg
func Do(quit <-chan struct{}, ...args) {
	select{
	case <-quit:
	} 
}
Do(ctx.Done(), args...)
```

I think explicit `quit` is better but i dont mind.

File: request/do.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply request/do.go:55]: `rec <- doRequest(url, timeout)` is already in a goroutine so it remains concurrent (otherwise it wouldn't have been working when the `for` wasn't wrapped in `func`), here this is just an anonymous func so we can `return`

File: cmd/runner/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply cmd/runner/main.go:57]: Yup, first one is really how context is meant to be used, the pattern is used everywhere and is one of the most Go-ish things. Passing `ctx.Done()` looks a bit unusual to me, but again, nitpick 🤷‍♂️

File: request/do.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/do.go:55]: Calling `func() {}()` (not a goroutine, what you suggest) is the same as just inlining the body. So i'll compare with that.


`Do` :
```go
	rec := make(chan Record, requests) // buffered, has enough capacity for the entire loop

	for i := 0; i < requests; i++ {
		select {
		case <-quit:
			wg.Wait()
			close(rec)
			return collect(rec) // not called until quit is received. rec is not processed until then
		default:
		}
		acquire(sem, &wg)
		go func() {
			defer release(sem, &wg)
			rec <- doRequest(url, timeout)
		}()
	} // blocks here

	return collect(rec) // not called until the loop is exited. rec is not processed until then
```

`DoUntil`:
```go
	rec := make(chan Record) // non-buffered, will block until someone receives from it

	for {
		select {
		case <-quit:
			wg.Wait()
			close(rec)
			return collect(rec) // not called until quit is received. This will never happen because we'll block at doRequest() (unless quit is receive JUST after the first iteration but before the second).
		default:
		}
		acquire(sem, &wg)
		go func() {
			defer release(sem, &wg)
			rec <- doRequest(url, timeout) // blocks until someone received from rec (ie second iteration)
		}()
	}
```

[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply request/do.go:55]: Ok, several things:

- I didn't notice you moved `wg.Wait()` and `close(rec)` inside `case <- quit`: this cannot work in `Do` as it'll never end until the global timeout is reached (we want to stop as soon as all requests are done). I think it should remain just before the final `return collect(rec)` and there should be only `return` in `case <- quit`

-  > Calling `func() {}()` (not a goroutine, what you suggest) is the same as just inlining the body
  Precisely, that is exactly what I meant: I expected the exact same behaviour as before (except the fixed `break` thing). That's why I suggested not to use a goroutine. Now try to use a goroutine after fixing the first point 🙃 

- I thought the `for` loop would be blocking only the time to throw the goroutines, not actually waiting for the request before going to the next iteration, this is indeed what we should aim for but I haven't figured why it isn't already the case 🤔

[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/do.go:55]: Fixed in 20c36dab9a0511ba106bee2232871e791a85f772.
The outer goroutine is still needed because the for loop (or forever loop) waits for a free threads before launching an inner goroutine for the request itself.

It's not n threads and do the calls, but ♾️ threads with at most n concurrent and do the calls.

[the remark this answers is not in the document]
Comment (GregoryAlbouy) [reply request/do.go:55]: Nothing's fixed 😬 It currently doesn't work.

Try running the following, the program never stops:
```sh
> go run cmd/runner/main.go -c 2 -r 10 http://localhost:9999\?delay\=500
```

[the remark this answers is not in the document]
Comment (moreirathomas) [reply request/do.go:55]: Works on my machine.

... just did not push

**Review by GregoryAlbouy — APPROVED** [review]

File: request/do.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+	return collect(rec)
+}
+
+// DoUntil launches a goroutine to ping url as soon as a thread is
+// available and collects the results as they come in.
+// The value of concurrency limits the number of concurrent threads.
+// On done signal from ctx, waits for goroutines to end and returns
+// the collected records.
+func DoUntil(quit context.Context, concurrency int, url string, timeout time.Duration) []Record {
```
Comment (GregoryAlbouy) [inline request/do.go:82]:
```suggestion
func DoUntil(ctx context.Context, concurrency int, url string, timeout time.Duration) []Record {
```
